### PR TITLE
added get_type_name and to_string to ParameterVariant

### PR DESCRIPTION
**Pull request by wjwwood**
This was done to enable more concise parameter examples.

I'm open to suggestion on a more elegant way to do the bytes to_string within a switch statement.

Connects to ros2/examples#4

### CODE REVIEW

File: rclcpp/include/rclcpp/parameter.hpp [hunk 1]
```diff
@@ -93,6 +95,30 @@ class ParameterVariant
 
   inline ParameterType get_type() const {return static_cast<ParameterType>(value_.type); }
 
+  inline std::string get_type_name() const
+  {
+    switch (get_type()) {
+      case rclcpp::parameter::ParameterType::PARAMETER_BOOL:
+        return "bool";
+        break;
```
Comment (dirk-thomas) [inline rclcpp/include/rclcpp/parameter.hpp]: These `break` statements are unreachable and should be removed.

File: rclcpp/include/rclcpp/parameter.hpp [hunk 2]
```diff
@@ -93,6 +95,30 @@ class ParameterVariant
...
+        return "double";
+        break;
+      case rclcpp::parameter::ParameterType::PARAMETER_STRING:
+        return "string";
+        break;
+      case rclcpp::parameter::ParameterType::PARAMETER_BYTES:
+        return "bytes";
+        break;
+      case rclcpp::parameter::ParameterType::PARAMETER_NOT_SET:
+      default:
```
Comment (dirk-thomas) [inline rclcpp/include/rclcpp/parameter.hpp]: The default case and the not-set case should be separate to catch the problem if new types are added in the future.

File: rclcpp/include/rclcpp/parameter.hpp [hunk 3]
```diff
@@ -193,11 +219,53 @@ class ParameterVariant
     return parameter;
   }
 
+  std::string to_string() const
+  {
+    std::stringstream bytes;
```
Comment (dirk-thomas) [inline rclcpp/include/rclcpp/parameter.hpp]: Since the stringstream is only used in one of the cases the variable should be moved to that scope.

Comment (wjwwood) [inline rclcpp/include/rclcpp/parameter.hpp]: You can't declare local variables inside of case statements.

File: rclcpp/include/rclcpp/parameter.hpp (hunk 2 above)
Comment (wjwwood) [inline rclcpp/include/rclcpp/parameter.hpp]: What should the default case do? Throw an exception?

File: rclcpp/include/rclcpp/parameter.hpp (hunk 3 above)
Comment (dirk-thomas) [inline rclcpp/include/rclcpp/parameter.hpp]: Sure you can. Just create a local scope with curly braces.

File: rclcpp/include/rclcpp/parameter.hpp (hunk 2 above)
Comment (dirk-thomas) [inline rclcpp/include/rclcpp/parameter.hpp]: I would suggest either `assert` or `throw`.

**Comment by esteve** [comment]
+1

File: rclcpp/include/rclcpp/parameter.hpp (hunk 3 above)
Comment (wjwwood) [inline rclcpp/include/rclcpp/parameter.hpp]: Yeah, that works. News to me.

**Comment by wjwwood** [comment]
Ok, I addressed the comments and squashed.

File: rclcpp/include/rclcpp/parameter.hpp [hunk 4]
```diff
@@ -93,6 +95,26 @@ class ParameterVariant
...
+      case rclcpp::parameter::ParameterType::PARAMETER_DOUBLE:
+        return "double";
+      case rclcpp::parameter::ParameterType::PARAMETER_STRING:
+        return "string";
+      case rclcpp::parameter::ParameterType::PARAMETER_BYTES:
+        return "bytes";
+      case rclcpp::parameter::ParameterType::PARAMETER_NOT_SET:
+        return "not set";
+      default:
+        throw std::runtime_error("Unexpected type from ParameterVariant");
```
Comment (dirk-thomas) [inline rclcpp/include/rclcpp/parameter.hpp]: Should the error string include the actual numeric value of `get_type()`?

File: rclcpp/include/rclcpp/parameter.hpp [hunk 5]
```diff
@@ -193,11 +215,51 @@ class ParameterVariant
...
+            {
+              bytes << "0x" << byte;
+              if (!first_byte) {
+                bytes << ", ";
+              } else {
+                first_byte = false;
+              }
+            }
+            return bytes.str();
+          }
```
Comment (dirk-thomas) [inline rclcpp/include/rclcpp/parameter.hpp]: This block should be two spaces less indented.

File: rclcpp/include/rclcpp/parameter.hpp (hunk 4 above)
Comment (wjwwood) [inline rclcpp/include/rclcpp/parameter.hpp]: I'm not sure that's very useful, but I'll include it.

File: rclcpp/include/rclcpp/parameter.hpp (hunk 5 above)
Comment (wjwwood) [inline rclcpp/include/rclcpp/parameter.hpp]: Good catch, done.

**Comment by dirk-thomas** [comment]
+1

File: rclcpp/include/rclcpp/parameter.hpp (hunk 2 above)
Comment (dirk-thomas) [inline rclcpp/include/rclcpp/parameter.hpp]: Separate another case where the same exception was used: 3b337fd98091bfb4853f81979c178db3b0b4f59a

File: rclcpp/include/rclcpp/parameter.hpp [hunk 6]
```diff
@@ -193,11 +216,52 @@ class ParameterVariant
...
+        return std::to_string(as_double());
+      case rclcpp::parameter::ParameterType::PARAMETER_STRING:
+        return as_string();
+      case rclcpp::parameter::ParameterType::PARAMETER_BYTES:
+        {
+          std::stringstream bytes;
+          bool first_byte = true;
+          bytes << "[" << std::hex;
+          for (auto & byte : as_bytes())
+          {
```
Comment (dirk-thomas) [inline rclcpp/include/rclcpp/parameter.hpp]: Style error: 3a4f14739ce0f0274d388cd5c40665fde3be9824

**Pull request merged**